### PR TITLE
vstart: support independent configuration and startup of rgw instances

### PR DESCRIPTION
**Pull request by kmroz**
RGW was a bit neglected in some of our scripting. Trying to remedy this.

This PR:
1. Allows restarting rgw independently of the cluster.
2. Tries to provide better support for multiple rgw instances (same
   cluster, or multi-cluster via mstart.sh) by providing independent config
   sections, etc.
3. Allows setting base port number on which to listen (instances will
   increment base port by instance number).
4. Changing base port number when re-starting overrides rgw_frontends
   allowing new instance to run on changed port.
5. Generates proper pid/log/asok/etc files in out/client.rgw.\* (needed to properly stop rgw instances via mstop.sh)

Fixes: http://tracker.ceph.com/issues/15226

### CODE REVIEW

File: src/vstart.sh [hunk 1]
```diff
@@ -617,6 +635,29 @@ EOF
...
+if [ "$start_rgw" -eq 1 ]; then
+    for rgw in `seq 0 $((CEPH_NUM_RGW-1))`
+    do
+	if [ "$new" -eq 1 ]; then
+            mkdir -p $CEPH_DEV_DIR/client.rgw.$rgw
+	    key_fn=$CEPH_DEV_DIR/client.rgw.$rgw/keyring
+	    if [ $overwrite_conf -eq 1 ]; then
+		    cat <<EOF >> $conf_fn
+[client.rgw.$rgw]
+        rgw frontends = fastcgi, civetweb port=$((CEPH_RGW_PORT+$rgw))
```
Comment (theanalyst) [inline src/vstart.sh]: should we drop fastcgi here?

Comment (kmroz) [inline src/vstart.sh]: @theanalyst - I wondered about this also, but ultimately left it as it was in the [global] section.

**Comment by theanalyst** [comment]
LGTM

**Comment by kmroz** [comment]
@theanalyst - Cool, thanks for reviewing.

**Comment by kmroz** [comment]
@liewegas / @yehudasa - Sage, Yehuda - Both of you guys put a lot into vstart.sh. If you have a moment, a review of this from you would be awesome. Thanks!

**Comment by kmroz** [comment]
Hey RGWers - Anything else needed to improve this PR?

**Comment by liewegas** [comment]
@yehudasa ping

**Comment by kmroz** [comment]
A gentle bump :)

**Comment by kmroz** [comment]
@yehudasa - ping++

**Comment by cbodley** [comment]
i gave this a try and was able to verify that the `--rgw_port` override stuff works as expected. one thing that didn't work, though, was increasing the `--rgw_num` after the initial run

this has to do with adding separate `[client.rgw.#]` sections to ceph.conf for each gateway. it seems like the `rgw_port_override` stuff would be a lot simpler if we didn't do this, and instead passed the `--rgw_frontends` argument explicitly on the command line. you can still use `--name client.rgw.#` to get separate log files. do you buy that?

**Comment by kmroz** [comment]
@cbodley - Yes, you're right. In my mind, `--rgw_num` is something that's relevant when provisioning the test cluster (ie. when `-n` is specified). My thinking was that manipulating the number of daemon instances would warrant tear down of the cluster, followed by re-provisioning. Anyway, I'm flexible. We can aim to have `--rgw_num` manipulate the number of instances even on restart if you think it's a worthwhile bit.

As to `--rgw-frontends`, I was hoping to abstract this from the user.

I will re-visit this PR more thoroughly, hopefully this week.

Thanks for the review!

**Comment by cbodley** [comment]
@kmroz agreed that other ceph daemons need to be provisioned for in the initial setup, but radosgw is more of a client than a daemon in this sense. you can add or remove gateways at runtime, and they'll cooperate through rados to share the storage

i doubt many people will need to add/remove gateways in a vstart cluster, but i think it would be nice to support as long as we don't sacrifice other features. which brings us to..

> As to `--rgw-frontends`, I was hoping to abstract this from the user.

by abstracting this, you mean having ceph.conf remember the port number so they don't have to specify a `--rgw_port` or `CEPH_RGW_PORT` next time?

i can see this being useful in a multisite setup; the gateway endpoints have to be included in the zone configuration, so you don't want the ports changing.. we've been using mstart.sh for multisite testing, which runs vstart.sh for multiple clusters and overrides `CEPH_RGW_PORT` for each (there's also src/test/rgw/test_multi.py that runs mstart.sh)

**Comment by yehudasa** [comment]
@kmroz @cbodley status of this PR?

**Comment by kmroz** [comment]
@yehudasa @cbodley - Sorry, was MIA for some time. Let me go over the notes of this PR again. As I recall, the major problem with my approach here (pointed out by Casey) was that I was treating RGW as a daemon (akin to MONs and OSDs) where really it is a Ceph client.

**Comment by dachary** [comment]
jenkins test this please (no logs)

**Comment by liewegas** [comment]
@kmroz ping

**Comment by kmroz** [comment]
@sage - I think this should be closed.  It's horribly outdated.

**Pull request closed without merging**